### PR TITLE
Auto-upgrade connectors when they are in use only a patch version update

### LINKED ISSUE #4470: Auto-upgrade connector versions when they have had backwards compatible updates
## Tell us about the problem you're trying to solve
I want my connection to pick up a backwards-compatible connector upgrade (bugfix or performance improvement) without any intervention from me

Specifically, this is during the upgrade process, it should upgrade not just unused connectors but also in-use connectors to the latest patch for the current minor version (as determined by that release's definition list).

### PR DESCRIPTION
**Pull request by marcosmarxm**
## What
For unused connectors Airbyte update connectors to latest version.
This code adds the option to upgrade connectors in use only when it's a patch version.

Closes https://github.com/airbytehq/airbyte/issues/4470

## How
Create a function to validate there is only a patch version update.
Apply the updating to in use connectors.

## Recommended reading order
1. `AirbyteVersion.java`
2. `DatabaseConfigPersistence.java`
3. `Tests`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java [hunk 1]
```diff
@@ -1937,6 +1949,16 @@ static boolean hasNewVersion(final String currentVersion, final String latestVer
     }
   }
 
+  @VisibleForTesting
+  static boolean hasNewPatchVersion(final String currentVersion, final String latestVersion) {
+    try {
+      return new AirbyteVersion(latestVersion).checkOnlyPatchVersionIsUpdated(new AirbyteVersion(currentVersion));
+    } catch (final Exception e) {
```
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1956]: what exception would be thrown? generally when we catch exceptions it is 1 of 2 special cases: 
1. there is a non-runtime exception we need to catch (in that case we want to catch that specific exception not the broader `Exception`)
2. we don't care if the code fails and always want to move on. 

I don't think this code meets either of these. I think you can just remove the try/catch.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java [hunk 2]
```diff
@@ -1859,7 +1859,19 @@ void updateConfigsFromSeed(final DSLContext ctx, final ConfigPersistence seedCon
...
+        if (hasNewPatchVersion(connectorInfo.dockerImageTag, latestImageTag)) {
+          // Update connector to the latest patch version
+          LOGGER.info("Connector {} needs update: {} vs {}", repository, connectorInfo.dockerImageTag, latestImageTag);
+          if (configType == ConfigSchema.STANDARD_SOURCE_DEFINITION) {
+            writeStandardSourceDefinition(Collections.singletonList(Jsons.object(latestDefinition, StandardSourceDefinition.class)), ctx);
+          } else if (configType == ConfigSchema.STANDARD_DESTINATION_DEFINITION) {
+            writeStandardDestinationDefinition(Collections.singletonList(Jsons.object(latestDefinition, StandardDestinationDefinition.class)), ctx);
+          } else {
+            throw new RuntimeException("Unknown config type " + configType);
+          }
```
Comment (cgardens) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1872]: nice! is this code the same as what's on lines 1908-1915. can we refactor it into a helper method? 

I actually think depending on how you write the helper, you could probably refactor 1895-1903. I think it is the same code with a different definition passed in as an arg.

Comment (cgardens) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1872]: it's very detail oriented code and so it is risky to have it repeated in 3 places.

**Review by cgardens — APPROVED** [review]
nice! I think there's one small refactor that's worth doing so that you leave this area nicer than when you found it. otherwise looks good!

**Review by tuliren — APPROVED** [review]

**Review by cgardens — APPROVED** [review]
the refactor looks great! agreed with @tuliren 's comments. then ship it!

File: airbyte-tests/src/automaticMigrationAcceptanceTest/java/io/airbyte/test/automaticMigrationAcceptance/MigrationAcceptanceTest.java [hunk 3]
```diff
@@ -238,7 +239,11 @@ private static void assertDestinationDefinitionInformation(final ApiClient apiCl
           foundPostgresDestinationDefinition = true;
         }
         case "8be1cf83-fde1-477f-a4ad-318d23c9f3c6" -> {
-          assertEquals("0.2.0", destinationDefinitionRead.getDockerImageTag());
+          final String tag = destinationDefinitionRead.getDockerImageTag();
+          final AirbyteVersion currentVersion = new AirbyteVersion(tag);
+          final AirbyteVersion previousVersion = new AirbyteVersion("0.2.0");
+          final AirbyteVersion finalVersion = (currentVersion.checkOnlyPatchVersionIsUpdatedComparedTo(previousVersion) ? currentVersion : previousVersion);
+          assertEquals(finalVersion.toString(), currentVersion.toString());
```
Comment (marcosmarxm) [inline airbyte-tests/src/automaticMigrationAcceptanceTest/java/io/airbyte/test/automaticMigrationAcceptance/MigrationAcceptanceTest.java:246]: @tuliren this PR implement an auto-upgrade for patch versions, this means for migration tests that run 0.17/0.32/dev the LocalCSV was having some problems.
The version the static file is `0.2.0` and latest version of this connector is `0.2.9` so when running the `dev` migration this test was failing.

any suggestions or this solution is good?

**Review by tuliren — APPROVED** [review]
The connector build failure is due to a flaky test. So your changes should be good to go!

**Pull request merged**